pgtransaction/tests/test_atomic.py

```diff
@@ -113,103 +114,148 @@ def test_atomic_retries_context_manager_not_allowed():
-            inner()
+            inner(attempts)
         except psycopg2.errors.SerializationFailure:
             pass
 
-    outer()
+    outer(attempts)
     assert 1 == Trade.objects.all().count()
-    assert 2 == dec.retry
+    assert len(attempts) == 1
 
 
 @pytest.mark.django_db(transaction=True)
 def test_atomic_retries_with_run_time_failure():
-    dec = atomic(isolation_level="REPEATABLE READ", retry=2)
-
     assert not Trade.objects.exists()
-    assert 2 == dec.retry
+    attempts = []
 
-    @dec
-    def outer():
+    @atomic(isolation_level="REPEATABLE READ", retry=2)
+    def outer(attempts):
+        attempts.append(True)
         ddf.G(Trade)
         raise RuntimeError
 
     with pytest.raises(RuntimeError):
-        outer()
+        outer(attempts)
 
     assert not Trade.objects.all().exists()
-    assert 2 == dec.retry  # We shouldn't retry on RuntimeErrors
+    assert len(attempts) == 1
 
 
 @pytest.mark.django_db(transaction=True)
 def test_atomic_retries_with_nested_atomic_and_outer_retry():
-    dec = atomic(isolation_level="REPEATABLE READ", retry=1)
-
-    assert 0 == Trade.objects.all().count()
-    assert 1 == dec.retry
+    assert not Trade.objects.exists()
+    attempts = []
 
-    @dec
-    def outer():
+    @atomic(isolation_level="REPEATABLE READ", retry=1)
+    def outer(attempts):
         ddf.G(Trade)
 
         @atomic
-        def inner():
+        def inner(attempts):
+            attempts.append(True)
             ddf.G(Trade)
 
-        inner()
-        if dec.retry == 1:
-            raise psycopg2.errors.SerializationFailure
+        inner(attempts)
+
+        if len(attempts) == 1:
+            raise OperationalError from psycopg2.errors.SerializationFailure
 
-    outer()
+    outer(attempts)
     assert 2 == Trade.objects.all().count()
-    assert 0 == dec.retry
+    assert len(attempts) == 2
+
+
+@pytest.mark.django_db(transaction=True)
+def test_concurrent_serialization_error():
+    """
+    Simulate a concurrency issue that will throw a serialization error.
+    Ensure that a retry is successful
+    """
+
+    def concurrent_update(barrier, trade, calls):
+
+        # We have to instantiate the decorator inside the function, otherwise
+        # it is shared among threads and causes the test to hang. It's uncertain
+        # what causes it to hang.
+        @atomic(isolation_level="SERIALIZABLE", retry=3)
+        def inner_update(trade, calls):
+            calls.append(True)
+            trade = Trade.objects.get(id=trade.id)
+            trade.price = 2
+            trade.save()
+            time.sleep(1)
+
+        barrier.wait()
+        inner_update(trade, calls)
+
+    barrier = threading.Barrier(2)
+    trade = ddf.G(Trade, price=1)
+    calls = []
+    t1 = threading.Thread(target=concurrent_update, args=[barrier, trade, calls])
+    t2 = threading.Thread(target=concurrent_update, args=[barrier, trade, calls])
+
+    t1.start()
+    t2.start()
+
+    t1.join()
+    t2.join()
+
```

Comment:
Really cool you got this to work. I tried pretty much exactly this set up minus the "barrier" thing (which I've never even heard of) and could not get things to work. I wonder barrier was the missing piece.

Reply:
Yea the barrier plus the small sleep time makes it pretty reliable